presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -294,6 +345,278 @@ public Page getOutput()
+                int newRemainingBytes = remainingBytes - rowSizes[i];
+
+                if (newRemainingBytes <= 0) {
+                    bufferFull = true;
+                    return max(i - startPosition, 1);
+                }
+                remainingBytes = newRemainingBytes;
+            }
+
+            return positionCount - startPosition;
+        }
+
+        private void flush(OutputBuffer outputBuffer)
+        {
+            if (isBufferEmpty()) {
+                return;
+            }
+
+            Slice slice = serialize();
+            SerializedPage serializedPage = serde.wrapBuffer(slice, bufferedRowCount);
+            outputBuffer.enqueue(lifespan, partition, Collections.singletonList(serializedPage));
+            pagesAdded.incrementAndGet();
+            rowsAdded.addAndGet(bufferedRowCount);
+            bufferedRowCount = 0;
+        }
+
+        // Copy the buffers in order into the final slice
+        private Slice serialize()
+        {
+            int bytes = getBuffersSizeInBytes();
+
+            Slice slice = Slices.wrappedBuffer(new byte[bytes]);
+            SliceOutput sliceOutput = slice.getOutput();
+            sliceOutput.writeInt(blockEncodingBuffers.length);  // channelCount
+
+            for (int i = 0; i < blockEncodingBuffers.length; i++) {
+                BlockEncodingBuffers currentBlockEncodingBuffers = blockEncodingBuffers[i];
+                currentBlockEncodingBuffers.writeTo(sliceOutput);
+            }
+            return sliceOutput.slice();
+        }
+
+        private void ensurePositionsCapacity()
+        {
+            if (positions.length <= positionCount) {
+                positions = Arrays.copyOf(positions, max(2 * positions.length, positionCount));
+            }
+        }
+
+        private void ensureRowSizesCapacity()
+        {
+            if (rowSizes.length <= positionCount) {
+                rowSizes = Arrays.copyOf(rowSizes, max(2 * rowSizes.length, positionCount));
+            }
+        }
+
+        private boolean isBufferEmpty()
+        {
+            return bufferedRowCount == 0;
+        }
+
+        private boolean isBufferFull()
+        {
+            return bufferFull;
+        }
+
+        private int getBuffersSizeInBytes()
+        {
+            int blockEncodingBuffersSize = 0;
+            for (int i = 0; i < blockEncodingBuffers.length; i++) {
+                BlockEncodingBuffers blockEncodingBuffers = this.blockEncodingBuffers[i];
+                blockEncodingBuffersSize += blockEncodingBuffers.getSizeInBytes();
+            }
+            return SIZE_OF_INT + blockEncodingBuffersSize;  // Channel count for one int
+        }
+    }
+
+    static class DecodedObjectNode
+    {
+        private final Object decodedBlock;
+        private final ArrayList<DecodedObjectNode> children;
+
+        public DecodedObjectNode(Object decodedBlock)
+        {
+            this.decodedBlock = decodedBlock;
```

Comment:
requireNonNull - `this.decodedBlock = requireNonNull(decodedBlock, "decodedBlock is null");`